File: 0901-1000/0968-Binary Tree Cameras/0968-Binary Tree Cameras.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int minCameraCover(TreeNode* root) {
        int total = 0;
        return (dfs(root, total) == 0 ? 1 : 0) + total;
    }

private:
    int dfs(TreeNode* root, int& total) {
        if (root == nullptr) {
            return -1;
        }

        int left = dfs(root->left, total);
        int right = dfs(root->right, total);
        if (left == 0 || right == 0) {
            ++total;
            return 1;
        }
        if (left == 1 || right == 1) {
            return 2;
        }
        else {
            return 0;
        }
    }
};

File: 0901-1000/0968-Binary Tree Cameras/0968-Binary Tree Cameras.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def minCameraCover(self, root: TreeNode) -> int:
        total = 0
        def dfs(node):
            if node is None:
                return -1

            left = dfs(node.left)
            right = dfs(node.right)
            if left == 0 or right == 0:
                nonlocal total
                total += 1
                return 1
            
            if left == 1 or right == 1:
                return 2
            else:
                return 0

        return (1 if dfs(root) == 0 else 0) + total
